Documentation/ExtensionArchitecture/Extbase/Reference/FileUpload.rst

```diff
@@ -0,0 +1,489 @@
+does not contain an invalid file extension. The API has support for custom
+validators, which can be created on demand.
+
+To avoid complexity and maintain data integrity, a file upload is only
+processed if the validation of all properties of a domain model is successful.
+In this first implementation, file uploads are not persisted/cached temporarily,
+so this means in any case of a validation failure ("normal" validators and file upload
+validation) a file upload must be performed again by users.
+
+Possible future enhancements of this functionality could enhance the existing
+`#[FileUpload]` attribute/annotation with configuration like a temporary storage
+location, or specifying additional custom validators (which can be done via the PHP-API as
+described below)
+
+..  _extbase_fileupload_attribute_configuration:
+File upload configuration with the `FileUpload` attribute
+---------------------------------------------------------
+
+File upload for a property of a domain model can be configured using the
+newly introduced :php:`\TYPO3\CMS\Extbase\Annotation\FileUpload` attribute.
+
+Example:
+
+..  code-block:: php
+
+    #[FileUpload([
+        'validation' => [
+            'required' => true,
+            'maxFiles' => 1,
+            'fileSize' => ['minimum' => '0K', 'maximum' => '2M'],
+            'allowedMimeTypes' => ['image/jpeg', 'image/png'],
+        ],
+        'uploadFolder' => '1:/user_upload/files/',
+    ])]
+    protected ?FileReference $file = null;
+
+All configuration settings of the
+:php:`\TYPO3\CMS\Extbase\Mvc\Controller\FileUploadConfiguration` object can
+be defined using the :php-short:`\TYPO3\CMS\Extbase\Annotation\FileUpload`
+attribute. It is however not possible
+to add custom validators using the
+:php-short:`\TYPO3\CMS\Extbase\Annotation\FileUpload` attribute, which you
+can achieve with a manual configuration as shown below.
+
+The currently available configuration array keys are:
+
+*   `validation` (:php:`array` with keys `required`, `maxFiles`, `minFiles`,
+    `fileSize`, `allowedMimeTypes`, `imageDimensions`, see
+    :ref:`extbase_fileupload_attribute-validationkeys`)
+*   `uploadFolder` (:php:`string`, destination folder)
+*   `duplicationBehavior` (:php:`object`, behaviour when file exists)
+*   `addRandomSuffix` (:php:`bool`, suffixing files)
+*   `createUploadFolderIfNotExist` (:php:`bool`, whether to create missing
+    directories)
+
+It is also possible to use the :php-short:`\TYPO3\CMS\Extbase\Annotation\FileUpload` annotation to configure
+file upload properties, but it is recommended to use the
+:php-short:`\TYPO3\CMS\Extbase\Annotation\FileUpload` attribute due to better readability.
+
+..  _extbase_fileupload_attribute-manual-configuration:
+
+Manual file upload configuration
+--------------------------------
+
+A file upload configuration can also be created manually and should be
+done in the :php:`initialize*Action`.
+
+Example:
+
+..  code-block:: php
+
```

Comment:
Please add a caption, so it is clear you are in a controller...

Reply:
done